src/java.base/share/classes/sun/nio/ch/ServerSocketChannelImpl.java

```diff
@@ -392,6 +392,7 @@ public SocketChannel accept() throws IOException {
 
         acceptLock.lock();
         try {
+            ensureOpen();
```

Comment:
Hello Mark, I think this may not be needed, especially since it's being called without synchronizing on `stateLock`. A couple of lines below this line, there's a call to `begin(blocking)` and the implementation of `begin()` already calls the `ensureOpen()` after acquiring a `stateLock` first.

Reply:
Please ignore these review comments for now. I might have mistaken what this change is proposing to do, partly because I missed some of the discussion in the JBS issue (the comments were auto-hidden and I now see the "load additional comments" option).

Reply:
> Please ignore these review comments for now. I might have mistaken what this change is proposing to do, partly because I missed some of the discussion in the JBS issue (the comments were auto-hidden and I now see the "load additional comments" option).

Right, there has been quite a bit of back and forth on this. The channel state needs to checked before "beginning" the I/O operation. If asynchronously closed while in the block then the more specific AsynchronousCloseException will be thrown.

Reply:
Do you think this additional `ensureOpen()` that we introduce here would need to be called while holding the `stateLock` monitor? I don't think it would be needed given the implementation of ensureOpen() checks a volatile field but then I see that existing calls to ensureOpen() are done while holding the monitor on `stateLock`.

I found it a bit odd that the existing calls to `ensureOpen()` are being invoked with a `stateLock` which as per its comment states that it is to be used for checking the `state` field, whereas the `ensureOpen()` itself just checks a separate `closed` field which is volatile.

Reply:
> Do you think this additional `ensureOpen()` that we introduce here would need to be called while holding the `stateLock` monitor? I don't think it would be needed given the implementation of ensureOpen() checks a volatile field but then I see that existing calls to ensureOpen() are done while holding the monitor on `stateLock`.

No, that would lead to contention when doing non-blocking I/O. The cases where you see ensureOpen called while holding the stateLock are blocking I/O ops or non-I/O ops that require more complex coordination with close.

Reply:
OK, thank you Alan for that detail.